### PR TITLE
Settings screen for password digit grouping

### PR DESCRIPTION
**Pull request by mattrubin**
**TL;DR:** This PR adds a settings screen that allows the user to choose whether they prefer passwords rendered in groups of two digits or groups of three.

----- 

This PR is the result of two separate issues I've been thinking about for a long time:
1. [Version 2.0](https://github.com/mattrubin/Authenticator/releases/tag/2.0.0) of the app changed how passwords are displayed: instead of showing the whole password as a single string of digits, the passwords are [chunked](https://en.wikipedia.org/wiki/Chunking_(psychology)) into two-digit groups to make them easier to read and remember. Since then, several users of the app [have reported](https://github.com/mattrubin/Authenticator/issues/214) that they have an easier time remembering six-digit passwords when they are chunked into groups of three digits. A [pull request](https://github.com/mattrubin/Authenticator/pull/216) from @beaucollins proposed to change the chunk size dynamically based on the length of the password, but despite the cleverness of this approach, it never quite felt like the right solution to me.
   I avoided making any changes on this for almost a year, for a very simple reason – I personally find it much easier to read and remember passwords chunked into groups of two, and the handful of users I had originally asked for input felt the same way. (At least one user wrote in an App Store review that the two-digit grouping is _the_ specific reason they use Authenticator over other two-factor apps.) I have been reluctant to make any changes when the current two-digit chunking strategy has been working for (anecdotally) most users.
2. As long as I have been developing Authenticator, I have tried to keep the app as simple as possible. I want the app to be clear and straightforward for users, and I didn't want to add a settings screen with a bunch of potentially-confusing options. In addition to complexity for users, adding user settings increases the number of different states in which the app operates, and can make testing and debugging significantly more complicated. For these reasons, I've avoided adding a settings screen to the app.

After thinking over the best way to present passwords that are readable and memorable for all users, I've concluded that the solution is to overcome my reluctance to add a settings screen to the app, and to add a toggle allowing the user to pick how they prefer for passwords to be grouped. Currently, the options are groups of two digits or groups of three digits. If I receive further user feedback on the issue, I will consider adding more options, but for now I want to keep things as simple as possible. While I think [the approach](https://github.com/mattrubin/Authenticator/pull/216) of changing the chunk size based on password length is aesthetically pleasing, I have found in my conversations with users that preferred chunk size for ease of memorization seems constant for a given person, regardless of password length.

I would love to hear what people think of this approach, and in particular to know whether this is is a satisfactory solution for those who prefer passwords in groups of three digits.

Screenshot with passwords in groups of two digits Screenshot of settings screen with toggle for password grouping size Screenshot with passwords in groups of three digits

### CODE REVIEW

**Comment by mide** [comment]
I'm not familiar with Swift, but this looks great! I'm very impressed by the options menu and how clear the two options are. Well done!

**Comment by mattrubin** [comment]
To clarify: all passwords will be chunked into the same size digit groups, regardless of password length, even if that results in the password ending with an incomplete group.

Screenshot with passwords in groups of two digits. A seven-digit password has three groups of two digits followed by a one-digit group. Screenshot with passwords in groups of three digits. A seven-digit password has two groups of three digits followed by a one-digit group. An eight-digit password has two groups of three digits followed by a two-digit group.

**Comment by beaucollins** [comment]
I understand the desire to not have a settings screen. I feel a trade off could be hiding the toggle behind the toolbar then revealing it by dragging down.

First launch sets the scroll offset to to show the option but future launches set it to the first password.

If you plan to have more options then the settings screen makes the most sense.

**Comment by mattrubin** [comment]
@beaucollins Putting the toggle at the top of the list and hiding it by default after first launch is clever, but adding the toggle to the main screen is probably too prominent a place for a control most users will only ever use one time (or zero times).

And, as you pointed out, a settings screen would eventually be made necessary by plenty of other potential features (TouchID, encrypted backups, etc.), so adding it now makes sense.

**Comment by mattrubin** [comment]
@mide @beaucollins Thanks for the feedback!
